node/cmd/ccq/utils.go

```diff
@@ -0,0 +1,231 @@
+			} else {
+				return nil, fmt.Errorf(`unsupported call type for API key "%s" in permissions file "%s"`, apiKey, fileName)
+			}
+
+			if _, exists := allowedCalls[callKey]; exists {
+				return nil, fmt.Errorf(`"%s" is a duplicate allowed call for API key "%s" in permissions file "%s"`, callKey, apiKey, fileName)
+			}
+
+			allowedCalls[callKey] = struct{}{}
+		}
+
+		pe := &permissionEntry{
+			userName:      user.UserName,
+			apiKey:        apiKey,
+			allowUnsigned: user.AllowUnsigned,
+			allowedCalls:  allowedCalls,
+		}
+
+		ret[apiKey] = pe
+	}
+
+	return ret, nil
+}
+
+// validateRequest verifies that this API key is allowed to do all of the calls in this request.
+func validateRequest(logger *zap.Logger, env common.Environment, perms Permissions, signerKey *ecdsa.PrivateKey, apiKey string, qr *gossipv1.SignedQueryRequest) error {
+	apiKey = strings.ToLower(apiKey)
+	permsForUser, exists := perms[strings.ToLower(apiKey)]
+	if !exists {
+		logger.Debug("invalid api key", zap.String("apiKey", apiKey))
+		return fmt.Errorf("invalid api key")
+	}
+
+	// TODO: Should we verify the signatures?
+
+	if len(qr.Signature) == 0 {
+		if !permsForUser.allowUnsigned || signerKey == nil {
+			logger.Debug("request not signed and unsigned requests not supported for apiKey",
+				zap.String("apiKey", apiKey),
+				zap.Bool("allowUnsigned", permsForUser.allowUnsigned),
+				zap.Bool("signerKeyConfigured", signerKey != nil),
+			)
+			return fmt.Errorf("request not signed")
+		}
+
+		// Sign the request using our key.
+		var err error
+		digest := query.QueryRequestDigest(env, qr.QueryRequest)
+		qr.Signature, err = ethCrypto.Sign(digest.Bytes(), signerKey)
+		if err != nil {
+			logger.Debug("failed to sign request", zap.String("apiKey", apiKey), zap.Error(err))
+			return fmt.Errorf("failed to sign request: %w", err)
+		}
+	}
+
+	var queryRequest query.QueryRequest
+	err := queryRequest.Unmarshal(qr.QueryRequest)
+	if err != nil {
+		logger.Debug("failed to unmarshal request", zap.String("apiKey", apiKey), zap.Error(err))
+		return fmt.Errorf("failed to unmarshal request: %w", err)
+	}
+
+	// Make sure the overall query request is sane.
+	if err := queryRequest.Validate(); err != nil {
+		logger.Debug("failed to validate request", zap.String("apiKey", apiKey), zap.Error(err))
+		return fmt.Errorf("failed to validate request: %w", err)
+	}
+
+	// Make sure they are allowed to make all of the calls that they are asking for.
+	for _, pcq := range queryRequest.PerChainQueries {
+		switch q := pcq.Query.(type) {
+		case *query.EthCallQueryRequest:
+			for _, callData := range q.CallData {
+				contractAddress, err := vaa.BytesToAddress(callData.To)
+				if err != nil {
+					logger.Debug("failed to parse contract address", zap.String("apiKey", apiKey), zap.String("contract", hex.EncodeToString(callData.To)), zap.Error(err))
+					return fmt.Errorf("failed to parse contract address: %w", err)
+				}
+				if len(callData.Data) < 4 {
```

Comment:
consider using a constant such as this for the few places where you validate this:

```go
const ETH_SELECTOR_LENGTH = 4
```